### PR TITLE
Add locset expressions for proximal and distal translation of locations

### LINKED ISSUE #1662: Feature: proximal and distal translation of locsets
We have a `(proximal n locset)` function that returns the _region_ formed by extending `n micrometers` proximally from the input locations. Same for `(distal n locset)`.

Add new `(proximal-translate n locset)` and `(distal-translate n locset)` functions that translate the locations in the input locset `n micrometers` proximally and distally respectively.

- [ ] C++ implementation
- [ ] unit tests
- [ ] documentation

### PR DESCRIPTION
**Pull request by bcumming**
Add `proximal-translate` and `distal-translate` locset expressions.

Fixes #1662

### CODE REVIEW

File: arbor/morph/locset.cpp [hunk 1]
```diff
@@ -116,6 +117,138 @@ std::ostream& operator<<(std::ostream& o, const terminal_& x) {
...
+    auto start = thingify(dt.start, p);
+    auto distance = dt.distance;
+
+    struct branch_interval {
+        msize_t bid;
+        double distance;
+    };
+
+    for (auto c: start) {
+        std::stack<branch_interval> branches_reached;
```
Comment (noraabiakar) [inline arbor/morph/locset.cpp:199]: Does this need to be a stack? It doesn't hurt, but we just sort the generated locsets in the end.

File: doc/concepts/labels.rst [hunk 2]
```diff
@@ -293,6 +293,36 @@ Locset expressions
...
+
+      The proximal translation of the terminal locations (left) a distance of 10 μm using the
+      expression ``(proximal-translate (terminal) 10)``.
+
+.. label:: (distal-translate ls:locset distance:real)
+
+    The set of locations that correspond to moving each location in ``ls`` in the distal direction
+    ``distance`` μm or to a terminal location, whichever is closest.
+    The number of locations in the output is greater than or equal the number in the input, with
+    multiple locations generated from input locations that are less than ``distance`` μm
```
Comment (noraabiakar) [inline doc/concepts/labels.rst:313]: more than `distance`?

**Review by noraabiakar — CHANGES_REQUESTED** [review]
Looks good, thanks. I have a question and a small correction in the docs.

File: arbor/morph/locset.cpp (hunk 1 above)
Comment (bcumming) [reply arbor/morph/locset.cpp:199]: The algorithm effectively uses a depth-first search to traverse the sub-trees, for which a stack is the natural fit.
The actual branch sections in the output are stored in a different data structure: the `std::vector` `L`.

File: doc/concepts/labels.rst (hunk 2 above)
Comment (bcumming) [reply doc/concepts/labels.rst:313]: I think I mean less, but might have to reword it to make sense.

Comment (noraabiakar) [reply doc/concepts/labels.rst:313]: Ah I actually see it now.
